File: arrays/intervals/my_calendar_II.rb
class Node
    attr_accessor :s, :e, :left, :right, :single_overlap

    def initialize(s, e)
        @s = s
        @e = e
        @left = nil
        @right = nil
        @single_overlap = false
    end
end

class MyCalendarTwo
    def initialize()
        @root = nil
    end

    def book(s, e)
        return false if !insertable?(s, e, @root)

        insert(s, e, @root)
        true
    end

    def insertable?(s, e, insert_root)
        return true if !insert_root
        return true if s >= e

        if e <= insert_root.s
            insertable?(s, e, insert_root.left)
        elsif s >= insert_root.e
            insertable?(s, e, insert_root.right)
        else # Overlap
            if insert_root.single_overlap
                false
            elsif s >= insert_root.s && e <= insert_root.e
                true
            else
                insertable?(s, insert_root.s, insert_root.left) && insertable?(insert_root.e, e, insert_root.right)
            end
        end
    end

    private
    # O(logH) solution, where H is the height of BT - we can improve further if we use balanced binary trees
    def insert(s, e, insert_root)
        node = Node.new(s, e)
        @root = node if !@root
        return node if !insert_root

        if s >= insert_root.e
            insert_root.right = insert(s, e, insert_root.right)
        elsif e <= insert_root.s
            insert_root.left = insert(s, e, insert_root.left)
        else # Overlap
            insert_root.single_overlap = true
            a = [insert_root.s, s].min
            b = [insert_root.s, s].max
            c = [insert_root.e, e].min
            d = [insert_root.e, e].max

            insert_root.s, insert_root.e = b, c
            insert_root.left, insert_root.right = insert(a, b, insert_root.left), insert(c, d, insert_root.right)
        end

        insert_root
    end
end


require 'test/unit'
extend Test::Unit::Assertions

calendar = MyCalendarTwo.new()
assert_equal(calendar.book(10, 20), true)
assert_equal(calendar.book(50, 60), true)
assert_equal(calendar.book(10, 40), true)
assert_equal(calendar.book(5, 15), false)
assert_equal(calendar.book(5, 10), true)
assert_equal(calendar.book(25, 55), true)

File: arrays/intervals/my_calendar_II.py
class Node:
    def __init__(self, start, end):
        self.start, self.end = start, end
        self.left, self.right = None, None
        self.single_overlap = False


class MyCalendarTwo:
    def __init__(self):
        self.root = None

    def book(self, start: int, end: int) -> bool:
        if not self.root:
            self.root = Node(start, end)
            return True

        if not self._can_insert(start, end, self.root): return False
        self._insert(start, end, self.root)
        return True

    def _can_insert(self, start, end, insert_root):
        if not insert_root: return True
        if start >= end: return True

        if end <= insert_root.start:   return self._can_insert(start, end, insert_root.left)    # Falls to the left
        elif start >= insert_root.end: return self._can_insert(start, end, insert_root.right)   # Falls to the right
        else:                                                                                   # Diff Types of Overlap
            if insert_root.single_overlap:                                  return False
            elif start >= insert_root.start and end <= insert_root.end:     return True
            else:
                a = min(insert_root.start, start)
                b = max(insert_root.start, start)
                c = min(insert_root.end, end)
                d = max(insert_root.end, end)
                return self._can_insert(a, b, insert_root.left) and self._can_insert(c, d, insert_root.right)

    def _insert(self, start, end, insert_root):
        if not insert_root:
            insert_root = Node(start, end)
            return insert_root
        if start >= end: return insert_root

        if start >= insert_root.end:   insert_root.right = self._insert(start, end, insert_root.right)
        elif end <= insert_root.start: insert_root.left = self._insert(start, end, insert_root.left)
        else:
            insert_root.single_overlap = True
            a = min(insert_root.start, start)
            b = max(insert_root.start, start)
            c = min(insert_root.end, end)
            d = max(insert_root.end, end)
            insert_root.start, insert_root.end = b, c
            insert_root.left, insert_root.right = self._insert(a, b, insert_root.left), self._insert(c, d, insert_root.right)

        return insert_root


# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)


# 731. My Calendar II
# https://leetcode.com/problems/my-calendar-ii/description/
